File: python/examples/new_apis.py
# SPDX-License-Identifier: Apache-2.0
"""
New APIs Example
Demonstrates:
  - Iterator.seek() — forward and reverse positional seek
  - KVStore/ColumnFamily.put_if_absent() — atomic conditional insert
  - KVStore/ColumnFamily.clear() — bulk truncation
  - KVStore/ColumnFamily.count() — entry count
  - KVStore.stats_reset() + extended stats() fields

Run:
    python examples/new_apis.py
"""

import os
import time

from snkv import KVStore, NotFoundError

DB_FILE = "new_apis_example.db"


# ---------------------------------------------------------------------------
# 1. Iterator seek — forward
# ---------------------------------------------------------------------------
def section_seek_forward(db: KVStore) -> None:
    print("\n--- 1. Iterator seek (forward) ---")

    db.put(b"apple",  b"fruit")
    db.put(b"banana", b"fruit")
    db.put(b"cherry", b"fruit")
    db.put(b"date",   b"fruit")

    with db.iterator() as it:
        # Seek to exact key.
        it.seek(b"banana")
        print(f"  seek(b'banana')  → {it.key!r}")

        # Seek between keys — lands on next key >= target.
        it.seek(b"bo")
        print(f"  seek(b'bo')      → {it.key!r}  (next >= target)")

        # Seek past last key → eof.
        it.seek(b"zzz")
        print(f"  seek(b'zzz')     → eof={it.eof}")

        # Seek then walk forward.
        it.seek(b"cherry")
        keys = []
        while not it.eof:
            keys.append(it.key.decode())
            it.next()
        print(f"  walk from b'cherry': {keys}")

    db.clear()


# ---------------------------------------------------------------------------
# 2. Iterator seek — reverse
# ---------------------------------------------------------------------------
def section_seek_reverse(db: KVStore) -> None:
    print("\n--- 2. Iterator seek (reverse) ---")

    db.put(b"aaa", b"v")
    db.put(b"bbb", b"v")
    db.put(b"ccc", b"v")

    with db.iterator(reverse=True) as it:
        it.last()

        # Seek to exact key.
        it.seek(b"bbb")
        print(f"  rev seek(b'bbb') → {it.key!r}")

        # Seek between keys — lands on nearest key <= target.
        it.seek(b"bbc")
        print(f"  rev seek(b'bbc') → {it.key!r}  (nearest <=)")

        # Seek before all keys → eof.
        it.seek(b"000")
        print(f"  rev seek(b'000') → eof={it.eof}")

    db.clear()


# ---------------------------------------------------------------------------
# 3. Iterator seek — prefix iterator
# ---------------------------------------------------------------------------
def section_seek_prefix(db: KVStore) -> None:
    print("\n--- 3. Seek within a prefix iterator ---")

    db.put(b"user:alice", b"1")
    db.put(b"user:bob",   b"2")
    db.put(b"user:carol", b"3")
    db.put(b"team:alpha", b"x")   # different prefix

    with db.iterator(prefix=b"user:") as it:
        it.seek(b"user:bob")
        print(f"  seek(b'user:bob') within prefix b'user:':")
        while not it.eof:
            print(f"    {it.key!r}")
            it.next()
    print("  (prefix boundary respected — b'team:alpha' not visited)")

    db.clear()


# ---------------------------------------------------------------------------
# 4. Seek chaining
# ---------------------------------------------------------------------------
def section_seek_chaining(db: KVStore) -> None:
    print("\n--- 4. Seek chaining ---")

    db.put(b"x", b"1")
    db.put(b"y", b"2")
    db.put(b"z", b"3")

    with db.iterator() as it:
        # seek() returns self, so calls can be chained.
        key = it.seek(b"y").key
        print(f"  it.seek(b'y').key = {key!r}")

    db.clear()


# ---------------------------------------------------------------------------
# 5. put_if_absent — default CF
# ---------------------------------------------------------------------------
def section_put_if_absent(db: KVStore) -> None:
    print("\n--- 5. put_if_absent (default CF) ---")

    # Key absent → inserted, returns True.
    inserted = db.put_if_absent(b"lock", b"owner:alice")
    print(f"  put_if_absent(b'lock', absent)  → {inserted}")

    # Key present → not inserted, returns False.
    inserted = db.put_if_absent(b"lock", b"owner:bob")
    print(f"  put_if_absent(b'lock', present) → {inserted}")

    print(f"  current value: {db.get(b'lock')!r}  (first write wins)")

    # put_if_absent with TTL — auto-releasing lock.
    inserted = db.put_if_absent(b"session:42", b"token-xyz", ttl=5)
    remaining = db.ttl(b"session:42")
    print(f"  put_if_absent(b'session:42', ttl=5s) → inserted={inserted}, "
          f"remaining={remaining:.2f}s")

    db.clear()


# ---------------------------------------------------------------------------
# 6. put_if_absent — named CF
# ---------------------------------------------------------------------------
def section_put_if_absent_cf(db: KVStore) -> None:
    print("\n--- 6. put_if_absent (named CF) ---")

    with db.create_column_family("dedup") as cf:
        # Deduplication: first write wins.
        print(f"  cf.put_if_absent(b'msg:001', first)  → {cf.put_if_absent(b'msg:001', b'hello')}")
        print(f"  cf.put_if_absent(b'msg:001', second) → {cf.put_if_absent(b'msg:001', b'world')}")
        print(f"  value: {cf.get(b'msg:001')!r}  (first write wins)")

        # Works correctly inside an explicit transaction.
        db.begin(write=True)
        cf.put_if_absent(b"msg:002", b"a")
        cf.put_if_absent(b"msg:002", b"b")   # same tx — should be blocked
        db.commit()
        print(f"  tx: msg:002 = {cf.get(b'msg:002')!r}")


# ---------------------------------------------------------------------------
# 7. clear
# ---------------------------------------------------------------------------
def section_clear(db: KVStore) -> None:
    print("\n--- 7. clear ---")

    for i in range(10):
        db.put(f"k{i}".encode(), b"v")

    print(f"  count before clear: {db.count()}")
    db.clear()
    print(f"  count after  clear: {db.count()}")

    # New inserts work normally after clear.
    db.put(b"fresh", b"start")
    print(f"  count after re-insert: {db.count()}")

    # CF-level clear is isolated from the default CF.
    with db.create_column_family("scratch") as cf:
        cf.put(b"tmp1", b"x")
        cf.put(b"tmp2", b"x")
        print(f"  CF \"scratch\" count before cf.clear(): {cf.count()}")
        cf.clear()
        print(f"  CF \"scratch\" count after  cf.clear(): {cf.count()}")

    print(f"  default CF count unchanged: {db.count()}")
    db.clear()


# ---------------------------------------------------------------------------
# 8. count
# ---------------------------------------------------------------------------
def section_count(db: KVStore) -> None:
    print("\n--- 8. count ---")

    print(f"  empty store: {db.count()}")

    for i in range(5):
        db.put(f"k{i}".encode(), b"v")
    print(f"  after 5 puts: {db.count()}")

    db.delete(b"k2")
    print(f"  after delete k2: {db.count()}")

    # count() includes expired-but-not-yet-purged keys.
    db._db.put_ttl(b"exp", b"v", int((time.time() - 1) * 1000))
    print(f"  with 1 expired (unpurged): {db.count()}")

    db.purge_expired()
    print(f"  after purge_expired: {db.count()}")

    # CF count is independent of default CF.
    with db.create_column_family("myCF") as cf:
        cf.put(b"a", b"v")
        cf.put(b"b", b"v")
        print(f"  CF count={cf.count()}, default CF count={db.count()} (isolated)")

    db.clear()


# ---------------------------------------------------------------------------
# 9. Extended stats + stats_reset
# ---------------------------------------------------------------------------
def section_stats(db: KVStore) -> None:
    print("\n--- 9. Extended stats + stats_reset ---")

    db.stats_reset()

    # Write: key=5 bytes, value=10 bytes → 15 bytes written.
    db.put(b"hello", b"0123456789")

    # Read: value=10 bytes → 10 bytes read.
    db.get(b"hello")

    # Trigger lazy TTL expiry.
    db._db.put_ttl(b"gone", b"x", int((time.time() - 1) * 1000))
    db.get(b"gone")   # lazy-expires on read

    st = db.stats()

    print(f"  puts          = {st['puts']}")
    print(f"  gets          = {st['gets']}")
    print(f"  bytes_written = {st['bytes_written']}")
    print(f"  bytes_read    = {st['bytes_read']}")
    print(f"  wal_commits   = {st['wal_commits']}")
    print(f"  checkpoints   = {st['checkpoints']}")
    print(f"  ttl_expired   = {st['ttl_expired']}")
    print(f"  ttl_purged    = {st['ttl_purged']}")
    print(f"  db_pages      = {st['db_pages']}")

    # Reset clears cumulative counters; db_pages remains live.
    db.stats_reset()
    st = db.stats()
    print(f"  after reset: puts={st['puts']} gets={st['gets']} "
          f"bytes_written={st['bytes_written']}")
    print(f"  db_pages after reset = {st['db_pages']}  (always live)")

    db.clear()


# ---------------------------------------------------------------------------
# 10. Real-world: idempotent job scheduler using put_if_absent
# ---------------------------------------------------------------------------
class JobScheduler:
    """
    Distributed job deduplication using put_if_absent.

    Only the first worker that claims a job ID proceeds to execute it.
    The lock expires automatically via TTL so jobs can be retried if
    the worker crashes.
    """

    def __init__(self, db: KVStore, timeout_s: float = 30) -> None:
        self._db      = db
        self._timeout = timeout_s
        try:
            self._cf = db.open_column_family("jobs")
        except NotFoundError:
            self._cf = db.create_column_family("jobs")

    def try_claim(self, job_id: str, worker_id: str) -> bool:
        """Return True if this worker claimed the job, False if already claimed."""
        key = job_id.encode()
        return self._cf.put_if_absent(key, worker_id.encode(), ttl=self._timeout)

    def release(self, job_id: str) -> None:
        self._cf.delete(job_id.encode())

    def close(self) -> None:
        self._cf.close()


def section_job_scheduler(db: KVStore) -> None:
    print("\n--- 10. Real-world: idempotent job scheduler ---")

    sched = JobScheduler(db, timeout_s=10)

    # Worker A claims job-001.
    claimed_a = sched.try_claim("job-001", "worker-A")
    print(f"  worker-A claims job-001: {claimed_a}")

    # Worker B tries the same job — should be blocked.
    claimed_b = sched.try_claim("job-001", "worker-B")
    print(f"  worker-B claims job-001: {claimed_b}  (blocked — already claimed)")

    # Worker A claims a different job.
    claimed_a2 = sched.try_claim("job-002", "worker-A")
    print(f"  worker-A claims job-002: {claimed_a2}  (different job, succeeds)")

    # After releasing, another worker can claim it.
    sched.release("job-001")
    reclaimed = sched.try_claim("job-001", "worker-B")
    print(f"  worker-B re-claims job-001 after release: {reclaimed}")

    sched.close()


# ---------------------------------------------------------------------------
def main() -> None:
    with KVStore(DB_FILE) as db:
        print("=== SNKV New APIs Example ===")

        section_seek_forward(db)
        section_seek_reverse(db)
        section_seek_prefix(db)
        section_seek_chaining(db)
        section_put_if_absent(db)
        section_put_if_absent_cf(db)
        section_clear(db)
        section_count(db)
        section_stats(db)
        section_job_scheduler(db)


if __name__ == "__main__":
    main()

    for ext in ("", "-wal", "-shm"):
        try:
            os.remove(DB_FILE + ext)
        except FileNotFoundError:
            pass

    print("\n[OK] new_apis.py example complete.")

File: examples/new_apis.c
/* SPDX-License-Identifier: Apache-2.0 */
/*
** New APIs Example
**
** Demonstrates:
**   - kvstore_iterator_seek() — forward and reverse positional seek
**   - kvstore_put_if_absent() / kvstore_cf_put_if_absent() — atomic conditional insert
**   - kvstore_clear() / kvstore_cf_clear() — bulk truncation
**   - kvstore_count() / kvstore_cf_count() — entry count
**   - kvstore_stats_reset() + extended kvstore_stats() fields
*/

#define SNKV_IMPLEMENTATION
#include "snkv.h"
#include <stdio.h>
#include <string.h>

/* ------------------------------------------------------------------ */
/* Section 1: Iterator seek — forward                                  */
/* ------------------------------------------------------------------ */
static void section_seek_forward(KVStore *pKV) {
    printf("\n--- 1. Iterator seek (forward) ---\n");

    kvstore_put(pKV, "apple",  5, "fruit",  5);
    kvstore_put(pKV, "banana", 6, "fruit",  5);
    kvstore_put(pKV, "cherry", 6, "fruit",  5);
    kvstore_put(pKV, "date",   4, "fruit",  5);

    KVIterator *pIter = NULL;
    kvstore_iterator_create(pKV, &pIter);

    /* Seek to exact key. */
    kvstore_iterator_seek(pIter, "banana", 6);
    if (!kvstore_iterator_eof(pIter)) {
        void *pKey; int nKey;
        kvstore_iterator_key(pIter, &pKey, &nKey);
        printf("  seek(\"banana\")  → \"%.*s\"\n", nKey, (char *)pKey);
    }

    /* Seek between keys — lands on next key >= target. */
    kvstore_iterator_seek(pIter, "bo", 2);
    if (!kvstore_iterator_eof(pIter)) {
        void *pKey; int nKey;
        kvstore_iterator_key(pIter, &pKey, &nKey);
        printf("  seek(\"bo\")      → \"%.*s\"  (next >= target)\n", nKey, (char *)pKey);
    }

    /* Seek past last key → eof. */
    kvstore_iterator_seek(pIter, "zzz", 3);
    printf("  seek(\"zzz\")     → eof=%d\n", kvstore_iterator_eof(pIter));

    /* Seek then walk forward. */
    kvstore_iterator_seek(pIter, "cherry", 6);
    printf("  walk from \"cherry\":");
    while (!kvstore_iterator_eof(pIter)) {
        void *pKey; int nKey;
        kvstore_iterator_key(pIter, &pKey, &nKey);
        printf(" %.*s", nKey, (char *)pKey);
        kvstore_iterator_next(pIter);
    }
    printf("\n");

    kvstore_iterator_close(pIter);
    kvstore_clear(pKV);
}

/* ------------------------------------------------------------------ */
/* Section 2: Iterator seek — reverse                                  */
/* ------------------------------------------------------------------ */
static void section_seek_reverse(KVStore *pKV) {
    printf("\n--- 2. Iterator seek (reverse) ---\n");

    kvstore_put(pKV, "aaa", 3, "v", 1);
    kvstore_put(pKV, "bbb", 3, "v", 1);
    kvstore_put(pKV, "ccc", 3, "v", 1);

    KVIterator *pIter = NULL;
    kvstore_reverse_iterator_create(pKV, &pIter);

    /* Seek on a reverse iterator → nearest key <= target. */
    kvstore_iterator_last(pIter);
    kvstore_iterator_seek(pIter, "bbb", 3);
    if (!kvstore_iterator_eof(pIter)) {
        void *pKey; int nKey;
        kvstore_iterator_key(pIter, &pKey, &nKey);
        printf("  rev seek(\"bbb\") → \"%.*s\"\n", nKey, (char *)pKey);
    }

    /* Seek between keys — lands on nearest key <= target. */
    kvstore_iterator_seek(pIter, "bbc", 3);
    if (!kvstore_iterator_eof(pIter)) {
        void *pKey; int nKey;
        kvstore_iterator_key(pIter, &pKey, &nKey);
        printf("  rev seek(\"bbc\") → \"%.*s\"  (nearest <=)\n", nKey, (char *)pKey);
    }

    /* Seek before all keys → eof. */
    kvstore_iterator_seek(pIter, "000", 3);
    printf("  rev seek(\"000\") → eof=%d\n", kvstore_iterator_eof(pIter));

    kvstore_iterator_close(pIter);
    kvstore_clear(pKV);
}

/* ------------------------------------------------------------------ */
/* Section 3: Iterator seek — prefix iterator                          */
/* ------------------------------------------------------------------ */
static void section_seek_prefix(KVStore *pKV) {
    printf("\n--- 3. Seek within a prefix iterator ---\n");

    kvstore_put(pKV, "user:alice", 10, "1", 1);
    kvstore_put(pKV, "user:bob",   8,  "2", 1);
    kvstore_put(pKV, "user:carol", 10, "3", 1);
    kvstore_put(pKV, "team:alpha", 10, "x", 1);  /* different prefix */

    KVIterator *pIter = NULL;
    kvstore_prefix_iterator_create(pKV, "user:", 5, &pIter);

    /* Seek to a key inside the prefix. */
    kvstore_iterator_seek(pIter, "user:bob", 8);
    printf("  seek(\"user:bob\") within prefix \"user:\":\n");
    while (!kvstore_iterator_eof(pIter)) {
        void *pKey; int nKey;
        kvstore_iterator_key(pIter, &pKey, &nKey);
        printf("    %.*s\n", nKey, (char *)pKey);
        kvstore_iterator_next(pIter);
    }
    printf("  (prefix boundary respected — \"team:alpha\" not visited)\n");

    kvstore_iterator_close(pIter);
    kvstore_clear(pKV);
}

/* ------------------------------------------------------------------ */
/* Section 4: put_if_absent — default CF                               */
/* ------------------------------------------------------------------ */
static void section_put_if_absent(KVStore *pKV) {
    printf("\n--- 4. put_if_absent (default CF) ---\n");

    int inserted = 0;

    /* Key absent → inserted. */
    kvstore_put_if_absent(pKV, "lock", 4, "owner:alice", 11, 0, &inserted);
    printf("  put_if_absent(\"lock\", absent) → inserted=%d\n", inserted);

    /* Key present → not inserted. */
    kvstore_put_if_absent(pKV, "lock", 4, "owner:bob", 9, 0, &inserted);
    printf("  put_if_absent(\"lock\", present) → inserted=%d\n", inserted);

    /* Verify the original value is untouched. */
    void *pVal = NULL; int nVal = 0;
    kvstore_get(pKV, "lock", 4, &pVal, &nVal);
    printf("  current value: \"%.*s\"\n", nVal, (char *)pVal);
    snkv_free(pVal);

    /* put_if_absent with TTL — insert a lock that auto-releases in 5 s. */
    kvstore_put_if_absent(pKV, "session:42", 10, "token-xyz", 9,
                          kvstore_now_ms() + 5000, &inserted);
    int64_t rem = 0;
    kvstore_ttl_remaining(pKV, "session:42", 10, &rem);
    printf("  put_if_absent(\"session:42\", ttl=5s) → inserted=%d, remaining=%lld ms\n",
           inserted, (long long)rem);

    kvstore_clear(pKV);
}

/* ------------------------------------------------------------------ */
/* Section 5: put_if_absent — named CF                                 */
/* ------------------------------------------------------------------ */
static void section_put_if_absent_cf(KVStore *pKV) {
    printf("\n--- 5. put_if_absent (named CF) ---\n");

    KVColumnFamily *pCF = NULL;
    kvstore_cf_create(pKV, "dedup", &pCF);

    int inserted = 0;

    kvstore_cf_put_if_absent(pCF, "msg:001", 7, "hello", 5, 0, &inserted);
    printf("  cf put_if_absent(\"msg:001\", first)  → inserted=%d\n", inserted);

    kvstore_cf_put_if_absent(pCF, "msg:001", 7, "world", 5, 0, &inserted);
    printf("  cf put_if_absent(\"msg:001\", second) → inserted=%d\n", inserted);

    void *pVal = NULL; int nVal = 0;
    kvstore_cf_get(pCF, "msg:001", 7, &pVal, &nVal);
    printf("  value: \"%.*s\"  (first write wins)\n", nVal, (char *)pVal);
    snkv_free(pVal);

    /* put_if_absent inside an explicit transaction. */
    kvstore_begin(pKV, 1);
    kvstore_cf_put_if_absent(pCF, "msg:002", 7, "a", 1, 0, &inserted);
    kvstore_cf_put_if_absent(pCF, "msg:002", 7, "b", 1, 0, &inserted);  /* same tx */
    kvstore_commit(pKV);

    kvstore_cf_get(pCF, "msg:002", 7, &pVal, &nVal);
    printf("  tx: msg:002 = \"%.*s\"  (first write in tx wins)\n", nVal, (char *)pVal);
    snkv_free(pVal);

    kvstore_cf_close(pCF);
}

/* ------------------------------------------------------------------ */
/* Section 6: clear                                                    */
/* ------------------------------------------------------------------ */
static void section_clear(KVStore *pKV) {
    printf("\n--- 6. clear ---\n");

    /* Populate default CF. */
    for (int i = 0; i < 10; i++) {
        char key[16]; int n = snprintf(key, sizeof key, "k%d", i);
        kvstore_put(pKV, key, n, "v", 1);
    }

    int64_t before = 0, after = 0;
    kvstore_count(pKV, &before);
    printf("  count before clear: %lld\n", (long long)before);

    kvstore_clear(pKV);
    kvstore_count(pKV, &after);
    printf("  count after  clear: %lld\n", (long long)after);

    /* New inserts work normally after clear. */
    kvstore_put(pKV, "fresh", 5, "start", 5);
    kvstore_count(pKV, &after);
    printf("  count after re-insert: %lld\n", (long long)after);

    /* CF-level clear is isolated. */
    KVColumnFamily *pCF = NULL;
    kvstore_cf_create(pKV, "scratch", &pCF);
    kvstore_cf_put(pCF, "tmp1", 4, "x", 1);
    kvstore_cf_put(pCF, "tmp2", 4, "x", 1);

    int64_t cfBefore = 0, cfAfter = 0;
    kvstore_cf_count(pCF, &cfBefore);
    kvstore_cf_clear(pCF);
    kvstore_cf_count(pCF, &cfAfter);
    printf("  CF \"scratch\": count %lld → %lld after cf_clear\n",
           (long long)cfBefore, (long long)cfAfter);

    /* Default CF unaffected. */
    int64_t defaultCount = 0;
    kvstore_count(pKV, &defaultCount);
    printf("  default CF count unchanged: %lld\n", (long long)defaultCount);

    kvstore_cf_close(pCF);
    kvstore_clear(pKV);
}

/* ------------------------------------------------------------------ */
/* Section 7: count                                                    */
/* ------------------------------------------------------------------ */
static void section_count(KVStore *pKV) {
    printf("\n--- 7. count ---\n");

    int64_t n = 0;
    kvstore_count(pKV, &n);
    printf("  empty store: count=%lld\n", (long long)n);

    for (int i = 0; i < 5; i++) {
        char key[8]; int kn = snprintf(key, sizeof key, "k%d", i);
        kvstore_put(pKV, key, kn, "v", 1);
    }
    kvstore_count(pKV, &n);
    printf("  after 5 puts: count=%lld\n", (long long)n);

    kvstore_delete(pKV, "k2", 2);
    kvstore_count(pKV, &n);
    printf("  after delete k2: count=%lld\n", (long long)n);

    /* count() includes expired-but-not-yet-purged keys. */
    int64_t past_ms = kvstore_now_ms() - 1000;
    kvstore_put_ttl(pKV, "exp", 3, "v", 1, past_ms);
    kvstore_count(pKV, &n);
    printf("  with 1 expired (unpurged): count=%lld\n", (long long)n);

    kvstore_purge_expired(pKV, NULL);
    kvstore_count(pKV, &n);
    printf("  after purge_expired: count=%lld\n", (long long)n);

    /* CF count is independent of default CF. */
    KVColumnFamily *pCF = NULL;
    kvstore_cf_create(pKV, "myCF", &pCF);
    kvstore_cf_put(pCF, "a", 1, "v", 1);
    kvstore_cf_put(pCF, "b", 1, "v", 1);

    int64_t cfN = 0;
    kvstore_cf_count(pCF, &cfN);
    int64_t defN = 0;
    kvstore_count(pKV, &defN);
    printf("  CF count=%lld, default CF count=%lld (isolated)\n",
           (long long)cfN, (long long)defN);

    kvstore_cf_close(pCF);
    kvstore_clear(pKV);
}

/* ------------------------------------------------------------------ */
/* Section 8: extended stats + stats_reset                             */
/* ------------------------------------------------------------------ */
static void section_stats(KVStore *pKV) {
    printf("\n--- 8. Extended stats + stats_reset ---\n");

    kvstore_stats_reset(pKV);

    /* Write a key. */
    kvstore_put(pKV, "hello", 5, "0123456789", 10);   /* 15 bytes */

    /* Read it back. */
    void *pVal = NULL; int nVal = 0;
    kvstore_get(pKV, "hello", 5, &pVal, &nVal);
    snkv_free(pVal);

    /* Add an expired TTL key and read it (triggers lazy expiry). */
    int64_t past_ms = kvstore_now_ms() - 1000;
    kvstore_put_ttl(pKV, "gone", 4, "x", 1, past_ms);
    { int64_t rem = 0;
      kvstore_get_ttl(pKV, "gone", 4, &pVal, &nVal, &rem); } /* lazy-expires on read */

    /* Purge nothing remaining, but record any purges. */
    kvstore_purge_expired(pKV, NULL);

    KVStoreStats st;
    kvstore_stats(pKV, &st);

    printf("  puts          = %llu\n",  (unsigned long long)st.nPuts);
    printf("  gets          = %llu\n",  (unsigned long long)st.nGets);
    printf("  bytes_written = %llu\n",  (unsigned long long)st.nBytesWritten);
    printf("  bytes_read    = %llu\n",  (unsigned long long)st.nBytesRead);
    printf("  wal_commits   = %llu\n",  (unsigned long long)st.nWalCommits);
    printf("  checkpoints   = %llu\n",  (unsigned long long)st.nCheckpoints);
    printf("  ttl_expired   = %llu\n",  (unsigned long long)st.nTtlExpired);
    printf("  ttl_purged    = %llu\n",  (unsigned long long)st.nTtlPurged);
    printf("  db_pages      = %llu\n",  (unsigned long long)st.nDbPages);

    /* Reset and verify cumulative counters clear. */
    kvstore_stats_reset(pKV);
    kvstore_stats(pKV, &st);
    printf("  after reset: puts=%llu gets=%llu bytes_written=%llu\n",
           (unsigned long long)st.nPuts,
           (unsigned long long)st.nGets,
           (unsigned long long)st.nBytesWritten);
    printf("  db_pages after reset = %llu  (always live)\n",
           (unsigned long long)st.nDbPages);

    kvstore_clear(pKV);
}

/* ------------------------------------------------------------------ */
int main(void) {
    const char *path = "new_apis_example.db";
    remove(path);

    KVStore *pKV = NULL;
    int rc = kvstore_open(path, &pKV, KVSTORE_JOURNAL_WAL);
    if (rc != KVSTORE_OK) {
        fprintf(stderr, "kvstore_open failed: %d\n", rc);
        return 1;
    }

    printf("=== SNKV New APIs Example ===");

    section_seek_forward(pKV);
    section_seek_reverse(pKV);
    section_seek_prefix(pKV);
    section_put_if_absent(pKV);
    section_put_if_absent_cf(pKV);
    section_clear(pKV);
    section_count(pKV);
    section_stats(pKV);

    kvstore_close(pKV);
    remove(path);
    remove("new_apis_example.db-wal");
    remove("new_apis_example.db-shm");

    printf("\n[OK] new_apis.c example complete.\n");
    return 0;
}
